documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+    // Vector2D<T1>? IVec2.GetTruncating<T1>();
+
+    // IReadOnlyList<T>
+    IEnumerator<T> IEnumerable<T>.GetEnumerator();
+    IEnumerator IEnumerable.GetEnumerator();
+    int IReadOnlyCollection<T>.Count => Count;
+
+    // IUtf8SpanFormattable, IUtf8SpanParsable<Vector2D<T>>
+    public bool TryFormat(Span<byte> utf8Destination, out int bytesWritten, ReadOnlySpan<char> format, IFormatProvider? provider);
+    public static Vector2D<T> Parse(ReadOnlySpan<byte> utf8Text, IFormatProvider? provider);
+    public static bool TryParse(ReadOnlySpan<byte> s, IFormatProvider? provider, out Vector2D<T> result);
+
+    // IVector<Vector2D<T>>
+    T IVector<Vector2D<T>, T>.LengthSquared();
+    static Vector2D<T> IVector<Vector2D<T>, T>.Multiply(Vector2D<T> left, Vector2D<T> right);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Multiply(Vector2D<T> left, T right);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Multiply(T left, Vector2D<T> right);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Negate(Vector2D<T> value);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Subtract(Vector2D<T> left, Vector2D<T> right);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Add(Vector2D<T> left, Vector2D<T> right);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Divide(Vector2D<T> left, Vector2D<T> right);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Divide(Vector2D<T> left, T divisor);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Clamp(Vector2D<T> value1, Vector2D<T> min, Vector2D<T> max);
+    static TReturn IVector<Vector2D<T>, T>.Distance<TReturn>(Vector2D<T> value1, Vector2D<T> value2);
+    static T IVector<Vector2D<T>, T>.DistanceSquared(Vector2D<T> value1, Vector2D<T> value2);
+    static TReturn IVector<Vector2D<T>, T>.DistanceSquared<TReturn>(Vector2D<T> value1, Vector2D<T> value2);
+    static T IVector<Vector2D<T>, T>.Dot(Vector2D<T> vector1, Vector2D<T> vector2);
+    static TReturn IVector<Vector2D<T>, T>.Dot<TReturn>(Vector2D<T> vector1, Vector2D<T> vector2);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Max(Vector2D<T> value1, Vector2D<T> value2);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Min(Vector2D<T> value1, Vector2D<T> value2);
+    static Vector2D<T> IVector<Vector2D<T>, T>.Lerp(Vector2D<T> value1, Vector2D<T> value2, T amount); /* where T : IFloatingPoint<T> */
+    static Vector2D<T> IVector<Vector2D<T>, T>.LerpClamped(Vector2D<T> value1, Vector2D<T> value2, T amount); /* where T : IFloatingPoint<T> */
+    static Vector2D<T> IVector<Vector2D<T>, T>.Reflect(Vector2D<T> vector, Vector2D<T> normal); /* where T : IFloatingPoint<T> */
+}
+
+public static Vector2D
+{
+    #region Extension
+    /// <summary>Returns the length of this vector object.</summary>
+    /// <returns>The vector's length.</returns>
+    /// <altmember cref="LengthSquared{T,TReturn}"/>
+    public static TReturn Length<T, TReturn>(this Vector2D<T> vec) where T : INumberBase<T> where TReturn : INumberBase<TReturn>, IRootFunctions<TReturn>;
```

Comment:
Having it differ like this is going to be incredibly difficult to get right.

Consider the difference between `Length<float, int>` and `Length<int, float>`  or even `Length<float, double>` and `Length<double, float>`. 

You effectively have to describe a behavior of how these operations function internally and when the conversion happens.

That is, is the behavior:
* Compute as T, then convert to TReturn and return
* Convert to TReturn, then compute and return

What happens matters significantly here since it impacts what combinations are valid, what intermediate computations may result, when overflow might occur, etc. Having the behavior differ based on the type of `T` or `TReturn` would significantly complicates things as well.

This is effectively why the Framework Design Guidelines say `T in, T out` and any conversions should be explicit by the user or statically lossless (such that an implicit conversion would have been viable).

Reply:
My original design was vec.Length<TReturn>(), but we made the call to use extension methods regardless of having to specify the generics again. More discussion needed.

Reply:
using an extension method with no TResult variant is my thought too but yeah more discussion needed